paasta_tools/cleanup_kubernetes_jobs.py

```diff
@@ -95,44 +104,118 @@ def alert_state_change(application: Application, soa_dir: str) -> Generator:
+            cluster=cluster, service=service, instance=instance, soa_dir=soa_dir
+        )
+        if len(applications) == 1:
+            application = applications[0]
+            if (
+                (service, instance) not in valid_services
+                or application.kube_deployment.namespace
+                != instance_config.get_namespace()
+            ):
+                applications_to_kill.append(application)
+        elif len(applications) > 1:
+            not_bouncing = instance_is_not_bouncing(
+                instance_config, applications, kube_client
+            )
+            for application in applications:
+                if (service, instance) not in valid_services or (
+                    application.kube_deployment.namespace
+                    != instance_config.get_namespace()
+                    and not_bouncing
+                ):
+                    applications_to_kill.append(application)
+    return applications_to_kill
+
+
 def cleanup_unused_apps(
-    soa_dir: str, kill_threshold: float = 0.5, force: bool = False
+    soa_dir: str,
+    cluster: str,
+    kill_threshold: float = 0.5,
+    force: bool = False,
 ) -> None:
     """Clean up old or invalid jobs/apps from kubernetes. Retrieves
     both a list of apps currently in kubernetes and a list of valid
     app ids in order to determine what to kill.
 
     :param soa_dir: The SOA config directory to read from
+    :param cluster: paasta cluster to clean
     :param kill_threshold: The decimal fraction of apps we think is
         sane to kill when this job runs.
     :param force: Force the cleanup if we are above the kill_threshold"""
     log.info("Creating KubeClient")
     kube_client = KubeClient()
 
     log.info("Loading running Kubernetes apps")
-    applications = list_namespaced_applications(
-        kube_client, "paasta", APPLICATION_TYPES
-    )
-
+    applications_dict = list_all_applications(kube_client, APPLICATION_TYPES)
     log.info("Retrieving valid apps from yelpsoa_configs")
     valid_services = set(
         get_services_for_cluster(instance_type="kubernetes", soa_dir=soa_dir)
     )
 
-    log.info("Determining apps to be killed")
-    applications_to_kill = [
-        applicaton
-        for applicaton in applications
-        if (applicaton.kube_deployment.service, applicaton.kube_deployment.instance)
-        not in valid_services
-    ]
-
-    log.debug("Running apps: %s" % applications)
+    applications_to_kill: List[Application] = get_applications_to_kill(
+        applications_dict, cluster, valid_services, kube_client, soa_dir
+    )
+
+    log.debug("Running apps: %s" % list(applications_dict))
     log.debug("Valid apps: %s" % valid_services)
     log.debug("Terminating: %s" % applications_to_kill)
     if applications_to_kill:
         above_kill_threshold = float(len(applications_to_kill)) / float(
-            len(applications)
+            len(list(applications_dict))
```

Comment:
out of curiosity: why are we converting this to a list? `len()` on a dictionary *should* return the # of keys

Reply:
TIL